Deprecate public APIs referencing Joda Time

headius opened the pull request:
This will be a series of commits to deprecate and replace all public APIs referencing classes from the Joda Time library and providing new versions based on java.time.

We will need to maintain these deprecated methods for a while before we can remove them and Joda Time from JRuby.

Some helpful guides:

* Mapping joda to java.time types, but not much help on how to convert: https://blog.joda.org/2014/11/converting-from-joda-time-to-javatime.html
* timeywimey a tiny library for converting some joda types to and from java.time: https://github.com/meetup/timeywimey

headius committed:
Deprecate public APIs referencing Joda Time

Joda Time is being deprecated in favor of the Java 8 java.time API
and we need to start moving away from the former. This patch marks
all public APIs that accept or return Joda DateTime as deprecated
and provides java.time.ZonedDateTime versions for migration. This
patch does NOT convert any internals to using java.time APIs.

There are other Joda classes referenced in public APIs. so this is
only part of the work needed for #6600.

I have not suppressed deprecation warnings so we can see where we
still call the deprecated public APIs in JRuby.

This makes use of the MIT-licensed "timeywimey" converted class,
see https://github.com/meetup/timeywimey.

headius commented:
@kares I think you may have looked into a full conversion in the past, perhaps you know how to handle some of the other APIs we have that use Joda types?

kares commented:
@headius great effort, what I ended up with with was pretty much the same as you - did not find direct replacement for JRuby's various use-cases of JODA `DateTime`. in some places it made sense to use ZonedTime but (e.g. in RubyTime although sometimes it's really a LocalTime we want), while others (RubyDate/RubyDateTime) seemed to need something different. 

have you looked at what TruffleRuby is using in the backend for the Ruby date/time types - they weren't using Java 8 time APIs back when I looked but now they should be?

headius commented:
@kares Last time I looked they were using the Java 8 time API and ZonedDateTime. Do you remember any specific examples of behavior that needed local time instead? I am not planning to carry this conversion forward for 9.3 but we'll need to do it at some point.

headius committed:
Deprecate remaining Joda APIs and fix warnings

This deprecates the remaining set of user-facing Joda Time-based
JRuby APIs:

* DateTime APIs are replaced with ZonedDateTime equivalents
* DateTimeZone APIs are replaced with ZoneId equivalents
* Chronology APIs are deprecated with no replacement

All internal uses of these deprecated APIs have either been moved
to the new java.time versions or supppress deprecation warnings.

headius commented:
I have pushed my last changes, which deprecate all remaining public-facing APIs with Joda classes and suppress or fix deprecation warnings internally to JRuby.

There will be some churn using the new java.time APIs since we still use Joda internally but this at least provides a migration path for third-party libraries. We can incrementally convert the internals to java.time in point releases leading up to a complete conversion by 9.4.

headius committed:
Zone handling tweaks

* Handle a null DateTimeZone as we did before, by propagating the
  null. This is used to indicate an invalid zone in our Ruby Time
  logic.
* Provide a map of zone aliases to pick up the old three-letter
  zones. This helps but is not sufficient to handle all
  three-letter formats tested by specs and tests (e.g. "PDT").
* For fixed-offset zones, use the offset rather than the ID. This
  causes those zones to lose their ID, so "AST-3:00:00" becomes
  an unnamed ZoneOffset of "3:00:00". This leads to other failures
  when the ID and offset are both expected to propagate into Time.

headius commented:
This is getting punted to 9.4, sadly.

The last couple commits led me to discover that java.time's time zone support is significantly stricter and narrower than the time zone support in Joda. It may not be possible to do a direct conversion from Joda to java.time without extending how java.time deals with custom or legacy time zone specifications. There may be time zone info providers out there for java.time that help fill in the gaps, but for now it is not possible to proceed with this conversion.

A couple examples:

* A TZ of "AST-3:00:00" should be treated as a time zone named "AST" with an offset of +3, but that does not correspond to any standard time zone ("AST" in java.time is Anchorage Standard Time at -10). As a result, we can only represent it in java.time as a ZoneOffset of +3 with no name.
* A TZ of "PST8:00:00" should be treated as a time zone named "PST" with an offset of -8, which matches Pacific Standard Time, but when parsed in this way it ends up as a nameless zone at -8.
* A TZ of "PDT" should be treated as Pacific Daylight Time but this ID does not roundtrip through java.time without error.

I believe now we will have to attempt the conversion from the inside out, matching behavior using java.time before we decide that we can ditch Joda. This won't happen for 9.3, and since we do not yet know whether we are able to use java.time to represent Ruby Time and Date, we cannot deprecate the Joda APIs.

headius commented:
I have rebased this on latest master but the problems remain: java.time is not a drop-in replacement due to strict and limited time zone support, so even if we deprecate the Joda Time APIs we can't start using java.time internally yet.

I find the java.time API to be inscrutable so I'm not sure how to proceed.

headius commented:
I'm going to close this for now. We have been unable to find a clean path that removes Joda from our codebase, given that the newer java.time API does not support all use cases we must maintain. If at some point we are able to figure out a clean path to using java.time, we can start deprecating this and providing alternative APIs. Deprecating without fully-functional alternatives does not help us.
